### PR TITLE
FEAT: Arrow Support in Bulk Copy

### LINKED ISSUE #37: Native Support for Arrow Bulk Copy
### Is your feature request related to a problem? Please describe.

The driver offers read support for Arrow datatypes. While it is feasible to bulkload arrow format by converting it to an iterator over tuples, I think if mssql-python took arrow input and serialized that over TDS, there may be a significant performance boost for arrow. 
Transforming data in place in memory and directly serializing to TDS can show significant performance improvement. A past experiement https://github.com/microsoft/mssql-rs/blob/dev/saurabh/arrow-benchmarks/mssql-arrow/BENCHMARK_REPORT.md shows that serialization from Arrow buffers to TDS without materializing the Rust types can be significantly faster compared to materializing to the Rust datatypes. 

### Describe the solution you'd like
Offer an API called which may look like the bulkcopy API but for arrow.

In line with the other arrow APIs in the driver, `bulkcopy_arrow(table, source...)` where source can be 
* ``pyarrow.Table``
* ``pyarrow.RecordBatch``
* ``pyarrow.RecordBatchReader``
* any object that exposes ``__arrow_c_stream__`` (Arrow PyCapsule
  interface)
* an iterable of ``pyarrow.RecordBatch`` (all batches must share
  the same schema)

### Describe alternatives you've considered
Arrow -> Iterator conversion -> BulkCopy

### PR DESCRIPTION
**Pull request by gargsaumya**
**Review Context:**

1. `mssql-py-core/src/arrow_bulkcopy.rs` - the type contract: build_column_plans → resolve_kind (one dispatch per column per batch) → extract_value per cell. Every unsupported (Arrow, SQL) pair fails fast before any row is written.
2. `mssql-py-core/src/cursor.rs` - the bulkcopy_arrow binding: GIL released for the transfer, re-acquired only to pump each batch; batch read/import errors are surfaced as a hard error.
3. `Python` side is a thin wrapper mirroring bulkcopy.

**In scope / supported destinations:** bit, all int/uint (range-checked narrowing; uint64→bigint overflow-checked), real/float, (n)char/(n)varchar/(n)text, binary/varbinary/image (incl. any-width fixed_size_binary), uniqueidentifier, date, datetime2, time, datetimeoffset, decimal/numeric, money/smallmoney, datetime/smalldatetime, xml, and json.

**Testing:** 37 Rust unit tests + 122 Arrow integration tests + a long-haul stress job wired into CI. 

## Description

This pull request introduces comprehensive support and test coverage for Arrow-based bulk copy operations in the `mssql-py-core` project, focusing on the integration of the Arrow library for high-performance data loading. It adds new pyarrow-based tests for various SQL types, updates pipeline templates for more flexible test selection and reporting, and ensures the Arrow dependency is included in both the Rust and Python environments.

**Arrow Bulk Copy Feature and Test Coverage**

* Added new test modules for Arrow-based bulk copy (`cursor.bulkcopy_arrow`) covering SQL types: BIGINT, BINARY/VARBINARY, BIT, and CHAR, including edge cases like null handling and type-specific constraints. [[1]](diffhunk://#diff-c788d77daf36d416ebe2fde7c856d279988dde24b241527b45cab93ff6f7b421R1-R146) [[2]](diffhunk://#diff-93cc8de741742d632e1a61238b8635866e575c4673f04005c391418f6452a78dR1-R168) [[3]](diffhunk://#diff-c8a77d6ededa0d2c25296e361411474b26f776e0572740c27161e758e8eefe24R1-R109) [[4]](diffhunk://#diff-71110f49cf27d0f7db9164623c097c5ea03aac90d510ef0dd0ea72d4a18fe46fR1-R101)
* Introduced the `arrow` crate as a dependency in `mssql-py-core/Cargo.toml` to enable Arrow integration in the Rust backend.
* Added the `arrow_bulkcopy` Rust module to the project structure.

**Pipeline and Test Infrastructure Improvements**

* Updated `.pipeline/templates/test-longhaul-template.yml` to parameterize test selection, results file, and test run title, allowing for more flexible and parallelized long-haul test execution. [[1]](diffhunk://#diff-c222b8ce091206a0f0f450e9fdc59dabbb17f678d32877a8f06083d36111ed46R15-R29) [[2]](diffhunk://#diff-c222b8ce091206a0f0f450e9fdc59dabbb17f678d32877a8f06083d36111ed46L102-R127) [[3]](diffhunk://#diff-c222b8ce091206a0f0f450e9fdc59dabbb17f678d32877a8f06083d36111ed46L127-R143)
* Modified `.pipeline/templates/validation-stages.yml` to run Arrow-specific long-haul tests in parallel with tuple-based tests, optimizing CI wall-clock time.
* Ensured `pyarrow` is installed in both pipeline and local development scripts (`dev/test-python.sh`).

These changes collectively improve the reliability, flexibility, and coverage of Arrow-based data loading in the project, and enhance the CI pipeline for future development.

## Related Issues
Fixes #37 and AB#46077

## Checklist

- [ ] `cargo bfmt` passes
- [ ] `cargo bclippy` passes
- [ ] `cargo btest` passes
- [ ] New/changed functionality has tests
- [ ] Public API changes are documented

### CODE REVIEW

File: mssql-py-core/src/arrow_bulkcopy.rs [hunk 1]
```diff
@@ -0,0 +1,1302 @@
...
+                scale: pick_timestamp_scale(dest.scale),
+            }
+        }
+
+        (DataType::Decimal128(_, s), S::Decimal | S::Numeric) => {
+            ColumnPlanKind::Decimal128 { scale: *s as u8 }
+        }
+
+        (DataType::FixedSizeBinary(16), S::UniqueIdentifier) => ColumnPlanKind::FixedBin16Uuid,
+        (DataType::FixedSizeBinary(16), S::Binary | S::VarBinary) => ColumnPlanKind::Binary,
```
Comment (David-Engel) [inline mssql-py-core/src/arrow_bulkcopy.rs]: This only maps `FixedSizeBinary(16)` to `BINARY`/`VARBINARY`, but valid Arrow fixed-size binary arrays can have other widths, for example `pa.binary(4)` for `BINARY(4)` or `VARBINARY(4)`. The extractor already handles `FixedSizeBinaryArray` generically below, so the planner rejects data the writer can serialize. Please keep the `FixedSizeBinary(16) -> UniqueIdentifier` arm, then allow `DataType::FixedSizeBinary(_)` for `S::Binary | S::VarBinary`, and add a fixed-size binary test; the current fixed-width BINARY test uses variable-width `pa.binary()` so it does not cover this.

**Review by David-Engel — COMMENTED** [review]
Can you improve the diff coverage to >= overall coverage? We rely heavily on test automation for validation so we want to maintain a high bar there.

**Review by David-Engel — COMMENTED** [review]
Better! One more push for your agent to cover more should do it. I'd suggest they make sure to include negative tests, based on the current missing lines.

File: mssql-py-core/src/cursor.rs [hunk 2]
```diff
@@ -491,6 +495,175 @@ impl PyCoreCursor {
...
+                    options.column_mappings
+                };
+
+                for mapping in mappings {
+                    let tds_mapping = match mapping {
+                        ColumnMapping::ByName {
+                            source,
+                            destination,
+                        } => TdsColumnMapping {
+                            source: ColumnMappingSource::Name(source),
```
Comment (saurabh500) [inline mssql-py-core/src/cursor.rs]: By-name source column mappings are accepted here at the Python layer (parse_one_tuple takes (str, str)), but resolve_column_mappings in mssql-tds/src/connection/bulk_copy.rs (~line 639) hard-rejects ColumnMappingSource::Name for iterator-based bulk copy with `Source column name mappings are not supported with iterator-based bulk copy`. So a caller who maps by Arrow field name -- the natural thing to do given Arrow tables carry a named schema -- passes validation at call time and then fails deep in resolution. Auto-mapping is also purely positional and ignores the Arrow schema names. Can we either support by-name source mapping for the Arrow path, or reject it up front at parse time with a clear message? Worth a follow-up issue to track this rather than blocking the PR.

File: mssql-py-core/src/arrow_bulkcopy.rs [hunk 3]
```diff
@@ -0,0 +1,1336 @@
...
+                unit: *unit,
+                scale: pick_timestamp_scale(dest.scale),
+            }
+        }
+
+        (DataType::Decimal128(_, s), S::Decimal | S::Numeric) => {
+            ColumnPlanKind::Decimal128 { scale: *s as u8 }
+        }
+
+        (DataType::FixedSizeBinary(16), S::UniqueIdentifier) => ColumnPlanKind::FixedBin16Uuid,
```
Comment (saurabh500) [inline mssql-py-core/src/arrow_bulkcopy.rs:256]: I ran a DB -> Arrow -> bulkcopy_arrow roundtrip against SQL Server 2022 across 25 types; 23 roundtripped cleanly. UNIQUEIDENTIFIER is the one genuine gap: mssql-python's cursor.arrow() reads a GUID column as Arrow large_string (the GUID text), but resolve_kind only accepts FixedSizeBinary(16) for a UniqueIdentifier dest, so the reload fails with `Cannot map Arrow column ... (LargeUtf8) to SQL column ... (UniqueIdentifier)`. That makes a real read-then-bulkload GUID roundtrip impossible today. Could we add a LargeUtf8/Utf8 -> UniqueIdentifier arm that parses the GUID string (accepting both binary and text forms)? If we'd rather keep the binary-only contract, let's document it explicitly. Suggest a follow-up issue to track. (For the record, DATETIMEOFFSET also mismatched in my test -- original +05:00 offset flattened to UTC and 100ns truncated to us -- but that's mssql-python's read normalization, not this PR.)

File: mssql-py-core/src/arrow_bulkcopy.rs [hunk 4]
```diff
@@ -0,0 +1,1336 @@
...
+        }
+        (DataType::Time64(unit @ (TimeUnit::Microsecond | TimeUnit::Nanosecond)), S::Time) => {
+            ColumnPlanKind::Time {
+                unit: *unit,
+                scale: pick_timestamp_scale(dest.scale),
+            }
+        }
+
+        (DataType::Decimal128(_, s), S::Decimal | S::Numeric) => {
+            ColumnPlanKind::Decimal128 { scale: *s as u8 }
```
Comment (saurabh500) [inline mssql-py-core/src/arrow_bulkcopy.rs]: Arrow's decimal scale is an i8 and is allowed to be negative (the Arrow spec permits it; data arriving over the C data interface from a non-pyarrow producer can carry a negative scale). `*s as u8` wraps a negative value into a large positive one (e.g. -2 becomes 254), which decimal128_to_string then mishandles and produces a garbage/overflowing string. The same `*s as u8` pattern is used in the Money and SmallMoney arms just below, so all three decimal paths share the bug. Suggest rejecting a negative scale up front with a clear error rather than silently wrapping:

``suggestion
        (DataType::Decimal128(_, s), S::Decimal | S::Numeric) => {
            if *s < 0 {
                return Err(Error::UsageError(format!(
                    "Negative Arrow decimal scale ({s}) is not supported"
                )));
            }
            ColumnPlanKind::Decimal128 { scale: *s as u8 }
        }
``

(and apply the same guard to the Money / SmallMoney arms). Worth a follow-up issue to track.

File: mssql-py-core/src/arrow_bulkcopy.rs [hunk 5]
```diff
@@ -0,0 +1,1336 @@
...
+                let v = downcast::<UInt64Array>(arr)?.value(row_idx);
+                if v > i64::MAX as u64 {
+                    return Err(Error::UsageError(format!(
+                        "Value {} out of range for BIGINT column '{}' (exceeds i64::MAX)",
+                        v, dest.column_name
+                    )));
+                }
+                narrow_int(v as i64, dest)
+            }
+            ColumnPlanKind::Float32 => match dest.sql_type {
```
Comment (saurabh500) [inline mssql-py-core/src/arrow_bulkcopy.rs]: Float32/Float64 values are wrapped straight into ColumnValues::Real/Float with no NaN/Infinity check. SQL Server rejects NaN and +/-Infinity for real/float columns, so today those cells fail with an opaque server-side error mid-stream (after some rows may already be written) rather than a clear client-side `NaN/Inf not allowed in column X`. Consider validating `value.is_finite()` here and returning a UsageError naming the column, so the failure is actionable and happens before we start writing. Also a good follow-up issue candidate.

**Comment by saurabh500** [comment]
@gargsaumya merge from main. Then the unit test reporting in mssql-py-core will start showing up as well in the diff coverage. That would help with the test numbers a lot

File: mssql-py-core/src/arrow_bulkcopy.rs (hunk 4 above)
Comment (gargsaumya) [reply mssql-py-core/src/arrow_bulkcopy.rs]: Good catch - fixed

File: mssql-py-core/src/arrow_bulkcopy.rs (hunk 5 above)
Comment (gargsaumya) [reply mssql-py-core/src/arrow_bulkcopy.rs]: Fixed.

File: mssql-py-core/src/arrow_bulkcopy.rs (hunk 3 above)
Comment (gargsaumya) [reply mssql-py-core/src/arrow_bulkcopy.rs:256]: Added.

File: mssql-py-core/src/cursor.rs (hunk 2 above)
Comment (gargsaumya) [reply mssql-py-core/src/cursor.rs]: Implemented by-name source mapping for the Arrow path. Since the Arrow schema carries field names, `bulkcopy_arrow` now resolves each `ByName` source to its ordinal against the schema before handoff, so the engine only sees ordinals — and an unknown name fails fast with `Arrow source column 'X' not found in schema (available: …)` instead of the deep iterator-based rejection. `column_mappings=[("arrow_field","DestCol"), …]` now works regardless of column order. Integration tests: `test_bulkcopy_arrow_by_name_source_mapping` (deliberately reordered columns) and `test_bulkcopy_arrow_by_name_unknown_source_errors`.

**Review by saurabh500 — APPROVED** [review]
